### ISSUE TITLE
Opened connections are not closed

### ISSUE THREAD

**Issue by abdoofathy95** [awslabs/aws-advanced-jdbc-wrapper#646]
### Describe the bug

Hey 👋 ,
I've started using this driver with a service that was originally using pgJDBC, I just replaced the driver with AWS JDBC driver. 

there is a behaviour that I'm seeing which is very consistent where opened connections keep increasing and never closed which causes a OOM.

I'm using Spring JDBC template to issue SQL queries to the DB.

### Expected Behavior

connections are closed and no OOM is thrown

### What plugins are used? What other connection properties were set?

auroraConnectionTracker,efm,failover,iam

### Current Behavior

org.postgresql.jdbc.PgConnection seem to not get closed

I have enabled driver logs and I can see this 
```
Opened Connections Tracked: [
[
	[ instance-1.rds.amazonaws.com:5432:
	 {
		org.postgresql.jdbc.PgConnection@245253d8,
		org.postgresql.jdbc.PgConnection@43a9541c
		.... 
         }]
]
``` 
which initially starts with connections = number of connections configured in the hikari pool (15), but after a couple of hours the number keeps increasing (reaching 100+) till a OOM is thrown.

this service have very little memory pressure yet it throws a out of heap space every couple of hours.

### Reproduction Steps

Using the following
```
AWS JDBC Version: 2.2.4
Postgresql runtime driver dependency: 42.6.0  
Spring boot: 2.7.11
pg-Aurora: 15.3 (2 read instances & 1 writer)
Hikari: 4.0.3
```

```
Driver config:
spring:
  datasource:
    type: com.zaxxer.hikari.HikariDataSource
    url: ${DB_URL}
    username: ${DB_USERNAME}
    driver-class-name: software.amazon.jdbc.Driver
    hikari:
      data-source-properties:
        wrapperPlugins: auroraConnectionTracker,efm,failover,iam
        wrapperDialect: aurora-pg
        wrapperLogUnclosedConnections: true
        wrapperLoggerLevel: ALL
      exception-override-class-name: software.amazon.jdbc.util.HikariCPSQLException
      max-lifetime: 300000
      connection-timeout: 1200
      validation-timeout: 1000
      maximum-pool-size: 15
```

and using the sslmode=verify-full & AWS RDS global-bundle.pem

### The AWS Advanced JDBC Driver version used

2.2.4

### JDK version used

oracle-jdk:17 (17.0.6+9-LTS-190)

### Operating System and version

Red Hat Enterprise Linux 8.7 (Ootpa)

**Comment by crystall-bitquill** [comment]
Hi @abdoofathy95,

Thanks for reaching out and raising this issue.

We'll take a look at this and keep you updated as we investigate.

In the meantime, could you provide the following additional information if possible?
- Driver logs, which can be enabled as seen [here](https://github.com/awslabs/aws-advanced-jdbc-wrapper/blob/main/docs/using-the-jdbc-driver/UsingTheJdbcDriver.md#logging).
- A bit of sample code, to get a better idea of how you're using the driver.

Thank you for your patience!

**Comment by abdoofathy95** [comment]
Hi @crystall-bitquill,
* for the driver logs, as mentioned above, the Opened Connections Tracked log shows a list that keeps growing (never shrinks), or are you interested in another bit of the log ? 
* a sample code:
```
spring:
  datasource:
    type: com.zaxxer.hikari.HikariDataSource
    url: ${DB_URL}
    username: ${DB_USERNAME}
    driver-class-name: software.amazon.jdbc.Driver
    hikari:
      data-source-properties:
        wrapperPlugins: auroraConnectionTracker,efm,failover,iam
        wrapperDialect: aurora-pg
        wrapperLogUnclosedConnections: true
        wrapperLoggerLevel: ALL
      exception-override-class-name: software.amazon.jdbc.util.HikariCPSQLException
      max-lifetime: 300000
      connection-timeout: 1200
      validation-timeout: 1000
      maximum-pool-size: 15
```

class Repository {
...
private final JdbcTemplate jdbcTemplate;
    public int[] query(paramArray) {
      return jdbcTemplate.batchUpdate("SQL", paramArray);
    }
}
....

**Comment by abdoofathy95** [comment]
Hey, this seems related mainly to using efm plugin, when disabled and instead using the following plugins 
`auroraConnectionTracker,failover,iam` not OOMs are observed. I didn't get the time to properly look at the exact root cause tho.

**Comment by crystall-bitquill** [comment]
Hi @abdoofathy95,

To update you, it seems like the additional connections you're seeing in the connection tracker comes from the Hikari pool. When idle connections reach the maxLifetime, they are closed and new connections are made. The connection tracker is logging these closed connections. However, all of these closed connections are weak references and can be removed by garbage collection. Have you made any changes to any garbage collection settings? If not, it is unlikely that the OOM error was caused by the connections visible in the tracker, as they should be garbage collected before the OOM error occurs. This occurs regardless of which plugins are used. 

You mentioned you didn't get the OOM error when efm was disabled. Would you be able to try the following plugin combinations to confirm where the error is?
1. No plugins - this can be done by setting the `wrapperPlugins` property to an empty string (the driver has several plugins enabled by default).
2. EFM only

In addition, I have a few questions:
1. Are you able to see how much free memory you have before and after starting your application?
2. Are you ablt to tell at what rate the amount of free memory you have available is decreasing?
3. Are any failover events occurring?

**Comment by abdoofathy95** [comment]
Hey @crystall-bitquill, 
thank you for your reply, 
I'm using the following flags
`-XX:+UseG1GC -Xms1536m -Xmx1536m -XX:MaxMetaspaceSize=256M -Xss2m`.

I will try to disable all the plugins and see the behaviour.

1. not sure I get what you mean here by before starting the application, it's a pod that as soon as it gets deployed, the application starts.
2. this is the JVM heap metric https://ibb.co/jLFJRHX
3. There are no failover events occurring.

**Comment by crystall-bitquill** [comment]
Hi @abdoofathy95,

To update you, similar issues (#675) have been reported regarding a memory leak in the Host Monitoring plugin.

We've recently added an **experimental** Host Monitoring Plugin (PR #764) to the driver, called the Host Monitoring Plugin v2. More information can be found [here](https://github.com/awslabs/aws-advanced-jdbc-wrapper/blob/main/docs/using-the-jdbc-driver/using-plugins/UsingTheHostMonitoringPlugin.md#experimental-host-monitoring-plugin-v2). The Host Monitoring Plugin v2 is functionally equivalent to the Host Monitoring Plugin and they are configured using the same parameters. The Host Monitoring Plugin v2 was created to address the issues that you and others have been experiencing, and it is meant to be a more stable version of the Host Monitoring Plugin.

As it is currently an experimental plugin, please note that it should be tested before being used in a production environment. This plugin can be tried out by using the snapshot build [`aws-advanced-jdbc-wrapper-2.3.2-20231213.213138-10.jar`](https://aws.oss.sonatype.org/content/repositories/snapshots/software/amazon/jdbc/aws-advanced-jdbc-wrapper/2.3.2-SNAPSHOT/). For more information on using a snapshot build, see [this page](https://github.com/awslabs/aws-advanced-jdbc-wrapper/blob/main/docs/using-the-jdbc-driver/UsingTheJdbcDriver.md#using-a-snapshot-of-the-driver). If you are able to test out the plugin, please let us know if you encounter any errors or if this plugin was able to resolve the issues you've encountered.

Thank you!

**Comment by aaronchung-bitquill** [comment]
Hi @abdoofathy95 ,

Heads up that the aforementioned experimental Host Monitoring Plugin is now available on our [latest release of the driver](https://search.maven.org/artifact/software.amazon.jdbc/aws-advanced-jdbc-wrapper/2.3.2/jar).

Also wanted to reiterate that it is currently an experimental plugin that should be tested before being used in a production environment.

If you are able to test out the plugin, please let us know if it resolves these issues or if there are any other issues encountered.

I'll be closing this ticket. But if this issue still persists or if you encounter any other issues, please feel free to create a new ticket or re-open this one.

Thank you
